dev/src/bulk-writer.ts

```diff
@@ -817,8 +855,54 @@ export class BulkWriter {
       this._errorFn.bind(this),
       this._successFn.bind(this)
     );
-    this._sendFn(enqueueOnBatchCallback, bulkWriterOp);
-    return bulkWriterOp.promise;
+
+    // Advance the `_lastOp` pointer. This ensures that `_lastOp` only resolves
+    // when both the previous and the current write resolves.
+    this._lastOp = this._lastOp.then(() => {
+      return silencePromise(bulkWriterOp.promise);
+    });
+
+    // Schedule the operation if the BulkWriter has fewer than the maximum
+    // number of allowed pending operations, or add the operation to the
+    // buffer.
+    if (this._pendingOpsCount < this._maxPendingOpCount) {
+      this._pendingOpsCount++;
+      this._sendFn(enqueueOnBatchCallback, bulkWriterOp);
+    } else {
+      this._bufferedOperations.push(() => {
+        this._pendingOpsCount++;
+        this._sendFn(enqueueOnBatchCallback, bulkWriterOp);
+      });
+    }
+
+    // Chain the BulkWriter operation promise with the buffer processing logic
+    // in order to ensure that it runs and that subsequent operations are
+    // enqueued before the next batch is scheduled in `_sendBatch()`.
+    return bulkWriterOp.promise
+      .then(res => {
+        this._processBufferedOps();
+        return res;
+      })
+      .catch(err => {
+        this._processBufferedOps();
+        throw err;
+      });
```

Comment:
You can probably return this._lastOp if you move this code to either line 862 above or _sendFn.

Reply:
I can't return lastOp since it's `Promise<void>`, but the return type is `Promise<WriteResult>`.